### PR TITLE
Upgrade to Terraform Plugin SDK v2

### LINKED ISSUE #159: Upgrade to Terraform Plugin SDK v2
We're currently using the "legacy Terraform plugin SDK" that's part of Terraform core. We'd like to switch to standalone, [Terraform Plugin SDK v2](https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html) to bring us up to date. A possibly contentious aspect of the upgrade is that we'd drop support for Terraform 0.11 and below. From my perspective this doesn't seem like an issue.

### PR DESCRIPTION
**Pull request by trotttrotttrott**
Minimal changes to migrate to [SDK v2](https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html). Closes https://github.com/grafana/terraform-provider-grafana/issues/159. Additional SDK v2 feature integrations will be done in a separate changeset (https://github.com/grafana/terraform-provider-grafana/issues/175).

[tf-sdk-migrator](https://github.com/hashicorp/tf-sdk-migrator) was primary used for code changes. Minor changes had to be made manually to fix acceptance tests.

**[Support for Terraform 0.11 and Below is being dropped](https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html#dropped-support-for-terraform-0-11-and-below)**. The readme now includes a note about a Terraform 0.12+ requirement.

This will not incur breaking changes for existing codebases (assuming use of Terraform 0.12+).

### CODE REVIEW

**Review by jvrplmlmn — APPROVED** [review]
Reviewed it commit by commit. Most changes are basically using the new imports and some minor fixes. Without having too much context about the provider, looks good to me 👍  Also neat that the upgrade already helped fixing some minor issues (drift 😄 ).

File: grafana/resource_dashboard_permission.go [hunk 1]
```diff
@@ -37,12 +37,12 @@ func ResourceDashboardPermission() *schema.Resource {
 						"team_id": {
 							Type:     schema.TypeInt,
 							Optional: true,
-							Default:  -1,
+							Default:  0,
```
Comment (inkel) [inline grafana/resource_dashboard_permission.go:40]: Just curious about the reason behind this change.

**Review by inkel — APPROVED** [review]

File: grafana/resource_dashboard_permission.go (hunk 1 above)
Comment (trotttrotttrott) [reply grafana/resource_dashboard_permission.go:40]: <details>
<summary>With SDKv2, the following acceptance test failure was occurring for both dashboard and folder permissions:</summary>

```
=== RUN   TestAccFolderPermission_basic
    resource_folder_permission_test.go:16: Step 1/2 error: After applying this test step and performing a `terraform refresh`, the plan was not empty.
        stdout


        An execution plan has been generated and is shown below.
        Resource actions are indicated with the following symbols:
          ~ update in-place

        Terraform will perform the following actions:

          # grafana_folder_permission.testPermission will be updated in-place
          ~ resource "grafana_folder_permission" "testPermission" {
                id         = "0XBNl58Gk"
                # (1 unchanged attribute hidden)

              + permissions {
                  + permission = "Admin"
                  + team_id    = -1
                  + user_id    = 3
                }
              - permissions {
                  - permission = "Admin" -> null
                  - team_id    = 0 -> null
                  - user_id    = 3 -> null
                }
              + permissions {
                  + permission = "Edit"
                  + role       = "Editor"
                  + team_id    = -1
                  + user_id    = -1
                }
              - permissions {
                  - permission = "Edit" -> null
                  - role       = "Editor" -> null
                  - team_id    = 0 -> null
                  - user_id    = 0 -> null
                }
              + permissions {
                  + permission = "View"
                  + team_id    = 2
                  + user_id    = -1
                }
              - permissions {
                  - permission = "View" -> null
                  - team_id    = 2 -> null
                  - user_id    = 0 -> null
                }
              + permissions {
                  + permission = "View"
                  + role       = "Viewer"
                  + team_id    = -1
                  + user_id    = -1
                }
              - permissions {
                  - permission = "View" -> null
                  - role       = "Viewer" -> null
                  - team_id    = 0 -> null
                  - user_id    = 0 -> null
                }
            }

        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccFolderPermission_basic (22.75s)
```

</details>

Terraform was writing `-1` for default values, but `terraform refresh` showed that `0` is what Grafana uses for unset foreign key values. `0` makes sense given it's the default for `int64` ([DashboardAclUpdateItem](https://github.com/grafana/grafana/blob/aee78f7527dbaf5e3a91f13504d2069229cddbfb/pkg/api/dtos/acl.go#L9-L14)).

It seems odd though that this did not cause drift without `terraform refresh`. Typically resources are refreshed as part of creating an execution plan. I suspect the answer is in the resources' read functions, but I'd have to investigate further.

Comment (inkel) [reply grafana/resource_dashboard_permission.go:40]: Thanks for explaining. Definitely odd there wasn't drift. I wonder if this wouldn't be masquerading some obscure issue down the line 🤔

**Pull request merged**